### ISSUE TITLE
FindAndReplace processor and ampersands

### ISSUE THREAD

**Issue by jelockwood** [autopkg/homebysix-recipes#601]
@homebysix 
I am trying to write an AutoPkg recipe to scrape a webpage which has a download link that is a dynamic URL with a time dependant token ID number. This URL also includes ampersands in it.

Here is an example of what the URL looks like when you view the page source -

```
<a href="//cdn.document360.io/e5d71abd-07b9-46d0-8876-03cc9073df6b/Images/Documentation/Jamf%20Compliance%20Editor%20v1.1.5.tar.gz?sv=2019-07-07&amp;sig=eJq3H6fTd%2BZ2OkpkmtvcVK%2BHcwjmyF3RmG8FD61CCDE%3D&amp;spr=https%2Chttp&amp;st=2023-07-31T16%3A15%3A33Z&amp;se=2023-07-31T16%3A25%3A33Z&amp;srt=o&amp;ss=b&amp;sp=r](http://cdn.document360.io/e5d71abd-07b9-46d0-8876-03cc9073df6b/Images/Documentation/Jamf%20Compliance%20Editor%20v1.1.5.tar.gz?sv=2019-07-07&amp;sig=eJq3H6fTd%2BZ2OkpkmtvcVK%2BHcwjmyF3RmG8FD61CCDE%3D&amp;spr=https%2Chttp&amp;st=2023-07-31T16%3A15%3A33Z&amp;se=2023-07-31T16%3A25%3A33Z&amp;srt=o&amp;ss=b&amp;sp=r)">Jamf Compliance Editor v1.1.5.tar.gz</a>
```
As you can see it contains ```%amp;``` to escape the **&** character.

I have a regex that successfully finds this URL on the full page but the resulting output string generated by the URLTextSearcher processor looks like the following

```
//cdn.document360.io/e5d71abd-07b9-46d0-8876-03cc9073df6b/Images/Documentation/Jamf%20Compliance%20Editor%20v1.1.5.tar.gz?sv=2019-07-07&amp;amp;sig=uMi62HCZg8XAcM33ADUS1WGZ3hTNqwuwBBaGp7Cyt50%3D&amp;amp;spr=https%2Chttp&amp;amp;st=2023-07-31T17%3A40%3A05Z&amp;amp;se=2023-07-31T17%3A50%3A05Z&amp;amp;srt=o&amp;amp;ss=b&amp;amp;sp=r](https://cdn.document360.io/e5d71abd-07b9-46d0-8876-03cc9073df6b/Images/Documentation/Jamf%20Compliance%20Editor%20v1.1.5.tar.gz?sv=2019-07-07&amp;amp;sig=uMi62HCZg8XAcM33ADUS1WGZ3hTNqwuwBBaGp7Cyt50%3D&amp;amp;spr=https%2Chttp&amp;amp;st=2023-07-31T17%3A40%3A05Z&amp;amp;se=2023-07-31T17%3A50%3A05Z&amp;amp;srt=o&amp;amp;ss=b&amp;amp;sp=r
```
If you look closely this results in the ```&amp;``` being converted to ```&amp;amp;```

This then results in the URLDownloader processor giving a 404 error.

Note: I have prepended the missing https: to the beginning so that is not the issue.

Testing confirms that the URL that actually needs to be used via CURL needs real **&** characters in it.

I therefore tried using your FindAndReplace processor to find &amp;amp; and replace that with a simple &, unfortunately this as shown below renders the recipe syntactically illegal.

```
    			<key>Arguments</key>
    			<dict>
        			<key>find</key>
        			<string>&amp;amp;</string>
        			<key>input_string</key>
        			<string>%match%</string>
        			<key>replace</key>
        			<string>&</string>
    			</dict>
    			<key>Processor</key>
   			<string>com.github.homebysix.FindAndReplace/FindAndReplace</string>
```

So, it looks like in the recipe I still need to escape the & but I cannot see how and still get a valid URL at the end of it. I have tried using your recipe to replace with ```&amp;``` and ```%26``` but neither of these worked when tested in either URLDownloader or a real web browser.

The following is what a valid version of the URL would look like.

```
https://cdn.document360.io/e5d71abd-07b9-46d0-8876-03cc9073df6b/Images/Documentation/Jamf%20Compliance%20Editor%20v1.1.5.tar.gz?sv=2019-07-07&sig=fVe6esbYaprSDk8%2FuDZV81%2FUcCyXTEvUWnthTiUndqY%3D&spr=https%2Chttp&st=2023-08-01T09%3A27%3A02Z&se=2023-08-01T09%3A37%3A02Z&srt=o&ss=b&sp=r
```

Any ideas?

**Comment by homebysix** [comment]
Could you try this?

```xml
				<key>find</key>
				<string>&amp;amp;</string>
				<key>replace</key>
				<string>&amp;</string>
```

**Comment by jelockwood** [comment]
@homebysix 
This is now resolved. Mainly because I found someone else had already written a download recipe for this piece of software and therefore all I had to do was write a Munki recipe. However having therefore also seen that persons download recipe which was mostly the same as my failed attempt I believe my error was not providing the 'web browser' agent details as part of my recipe.

PS. I had already tried the option you suggested but it did not work with my presumably bad recipe 😄
